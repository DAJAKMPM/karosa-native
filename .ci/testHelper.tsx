import React, { JSXElementConstructor } from "react";
import { View } from "react-native";
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from "@testing-library/react-native";
import { Formik, FormikValues } from "formik";

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };

  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || "wrapper");
  return { wrapper, ...renderApi };
};

export const renderWithFormikWrapper = <T extends FormikValues>(
  initialValues: T,
  children: React.ReactElement<any, string | JSXElementConstructor<any>>
) => {
  return renderWithWrapper(
    <Formik
      initialValues={initialValues}
      onSubmit={() => console.log("submit")}
      component={() => children}
    />
  );
};
