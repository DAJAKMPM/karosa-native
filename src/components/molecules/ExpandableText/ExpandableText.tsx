/**
 *
 * ExpandableText
 * @format
 *
 */

import React, { FC, useState, useEffect, useCallback } from "react";
import Text from "@app/atoms/Text";

import type { PropsType } from "./types";

const ExpandableText: FC<PropsType> = (props) => {
  const { readMoreStyle, text, textStyle } = props;

  const [textShown, setTextShown] = useState<boolean>(false);
  const [numLines, setNumLines] = useState<number | undefined>(undefined);

  const toggleTextShown = () => {
    setTextShown(!textShown);
  };

  useEffect(() => {
    setNumLines(textShown ? undefined : 4);
  }, [textShown]);

  const onTextLayout = useCallback(
    (e) => {
      if (e.nativeEvent.lines.length > 4 && !textShown) {
        setNumLines(4);
      }
    },
    [textShown]
  );

  return (
    <>
      <Text
        text={text}
        textStyle={textStyle}
        onTextLayout={onTextLayout}
        numberOfLines={numLines}
        ellipsizeMode="tail"
      />

      <Text
        text={textShown ? "Read Less" : "Read More"}
        textStyle={readMoreStyle}
        onPress={toggleTextShown}
      />
    </>
  );
};

export default ExpandableText;
