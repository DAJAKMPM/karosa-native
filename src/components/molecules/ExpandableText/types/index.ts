/**
 *
 * @format
 *
 */

import { StyleProp, TextStyle } from "react-native";

export type PropsType = {
  text: string;
  textStyle?: StyleProp<TextStyle>;
  readMoreStyle?: StyleProp<TextStyle>;
};
