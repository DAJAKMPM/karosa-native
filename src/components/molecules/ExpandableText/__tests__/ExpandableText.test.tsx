import React from "react";
import { renderWithWrapper } from "@app/ci/testHelper";
import ExpandableText from "../ExpandableText";

describe("ExpandableText Component", () => {
  it("should match snapshot", () => {
    const { toJSON } = renderWithWrapper(<ExpandableText text="test" />);
    expect(toJSON()).toMatchSnapshot();
  });
});
