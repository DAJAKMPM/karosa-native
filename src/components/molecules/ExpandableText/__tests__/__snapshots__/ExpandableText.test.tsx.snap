// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandableText Component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <Text
    ellipsizeMode="tail"
    numberOfLines={4}
    onTextLayout={[Function]}
    style={
      Object {
        "color": "#1C1C1C",
        "fontFamily": "SFProText-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 18.75,
      }
    }
  >
    test
  </Text>
  <Text
    onPress={[Function]}
    style={
      Object {
        "color": "#1C1C1C",
        "fontFamily": "SFProText-Regular",
        "fontSize": 16,
        "fontWeight": "400",
        "lineHeight": 18.75,
      }
    }
  >
    Read More
  </Text>
</View>
`;
